controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/AbstractRequestProcessor.java

```diff
@@ -0,0 +1,184 @@
+import io.pravega.shared.controller.event.DeleteStreamEvent;
+import io.pravega.shared.controller.event.RequestProcessor;
+import io.pravega.shared.controller.event.ScaleOpEvent;
+import io.pravega.shared.controller.event.SealStreamEvent;
+import io.pravega.shared.controller.event.TruncateStreamEvent;
+import io.pravega.shared.controller.event.UpdateStreamEvent;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import static io.pravega.controller.eventProcessor.impl.EventProcessorHelper.withRetries;
+
+/**
+ * Abstract common class for all request processing done over SerializedRequestHandler.
+ * This implements RequestProcessor interface and implements failing request processing for all ControllerEvent types with
+ * RequestUnsupported.
+ * Its derived classes should implement specific processing that they wish to handle.
+ *
+ * This class provides a common completion method which implements mechanisms that allow multiple event processors
+ * working on same stream to get fairness in their scheduling and avoid starvation.
+ * To do this, before starting any processing, it fetches waiting request processor record from the store and if there
+ * is a record in the store and it doesnt match current processing request, then it simply postpones the current processing.
+ * Otherwise it attempts to process the event. If the event fails in its processing because of contention with another
+ * process on a different processor, then it sets itself as the waiting request processor in the stream metadata. This will
+ * ensure that when other event processors complete their processing, they will not pick newer work until this processor
+ * processes at least one event.
+ * At the end of processing, each processor attempts to clean up waiting request processor record from the store if it
+ * was set against its name.
+ */
+@Slf4j
+public abstract class AbstractRequestProcessor<T extends ControllerEvent> extends SerializedRequestHandler<T> implements RequestProcessor {
+    protected static final Predicate<Throwable> OPERATION_NOT_ALLOWED_PREDICATE = e -> Exceptions.unwrap(e) instanceof StoreException.OperationNotAllowedException;
+
+    protected final StreamMetadataStore streamMetadataStore;
+
+    public AbstractRequestProcessor(StreamMetadataStore streamMetadataStore, ScheduledExecutorService executor) {
+        super(executor);
+        this.streamMetadataStore = streamMetadataStore;
+    }
+
+    abstract String getProcessorName();
+
+    @Override
+    public CompletableFuture<Void> processEvent(ControllerEvent controllerEvent) {
+        return controllerEvent.process(this);
+    }
+
+    @Override
+    public CompletableFuture<Void> processAbortTxnRequest(AbortEvent abortEvent) {
+        return Futures.failedFuture(new RequestUnsupportedException(
```

Comment:
Nit pick: don't these fit on one line? You have much longer lines above.

Reply:
Also, is there something special about `RequestUnsupportedException`? Would `UnsupportedOperationException` do?